File: app/src/main/java/com/nfcalarmclock/ratemyapp/NacRateMyApp.java
package com.nfcalarmclock.ratemyapp;

import androidx.appcompat.app.AppCompatActivity;
import com.google.android.play.core.review.ReviewInfo;
import com.google.android.play.core.review.ReviewManager;
import com.google.android.play.core.review.ReviewManagerFactory;
import com.google.android.play.core.tasks.Task;

import com.nfcalarmclock.shared.NacSharedPreferences;

/**
 * Handle when to prompt the user to rate my app.
 */
public final class NacRateMyApp
{

	/**
	 */
	private NacRateMyApp()
	{
	}

	/**
	 * Request to rate my app.
	 *
	 * If the app is already rated, or the counter has not reached the threshold
	 * to launch the request, then nothing is shown to the user.
	 */
	public static void request(NacSharedPreferences shared)
	{
		// App is already rated
		if (shared.isRateMyAppRated())
		{
			return;
		}

		// Time to show the Google rating dialog
		else if (shared.isRateMyAppLimit())
		{
			AppCompatActivity activity = (AppCompatActivity) shared.getContext();
			ReviewManager manager = ReviewManagerFactory.create(activity);
			Task<ReviewInfo> request = manager.requestReviewFlow();

			// Launch the review flow
			request.addOnCompleteListener(task -> {
				if (task.isSuccessful())
				{
					ReviewInfo reviewInfo = task.getResult();
					Task<Void> flow = manager.launchReviewFlow(activity, reviewInfo);

					// The flow has finished. The API does not indicate whether the user
					// reviewed or not, or even whether the review dialog was shown.
					// 
					// No matter the result, just treat it as rated.
					flow.addOnCompleteListener(ignored -> {
						shared.ratedRateMyApp();
					});
				}
			});

			//NacRateMyAppDialog dialog = new NacRateMyAppDialog();
			//dialog.build(this);
			//dialog.show();
		}

		// Increment the counter until it is time to show the Google rating dialog
		else
		{
			shared.incrementRateMyApp();
		}
	}

}

File: app/src/googleplay/kotlin/com/nfcalarmclock/ratemyapp/NacRateMyApp.kt
package com.nfcalarmclock.ratemyapp

import androidx.appcompat.app.AppCompatActivity
import com.google.android.play.core.review.ReviewInfo
import com.google.android.play.core.review.ReviewManagerFactory
import com.google.android.gms.tasks.Task
import com.nfcalarmclock.R
import com.nfcalarmclock.shared.NacSharedPreferences

/**
 * Handle when to prompt the user to rate my app.
 */
object NacRateMyApp
{

	/**
	 * Check if the rate my app Google Play flow should be shown or not.
	 *
	 * Note: If the flow should not be shown and the user has never been shown
	 *       the flow before, then the rate my app counter will be incremented.
	 *
	 * @return True if the rate my app Google Play flow should be shown, and
	 *         False otherwise.
	 */
	fun shouldRequest(shared: NacSharedPreferences): Boolean
	{
		// App is already rated
		return if (shared.isRateMyAppRated)
		{
			false
		}
		// Reached the counter limit so it is time to ask again
		else if (shared.isRateMyAppLimit)
		{
			true
		}
		// Increment the counter
		else
		{
			shared.rateMyAppCounter = shared.rateMyAppCounter + 1
			false
		}
	}

	/**
	 * Request to rate my app.
	 *
	 * This will NOT check if the review flow should be launched or not. It
	 * will simply launch it.
	 */
	fun request(activity:AppCompatActivity, shared: NacSharedPreferences)
	{
		// Create a review manager
		val manager = ReviewManagerFactory.create(activity)

		// Retrieves all the needed information to launch the review flow
		val request = manager.requestReviewFlow()

		// Launch the review flow
		request.addOnCompleteListener { task: Task<ReviewInfo> ->

			// Task was not successful
			if (!task.isSuccessful)
			{
				return@addOnCompleteListener
			}

			// Launch the review flow
			val flow = manager.launchReviewFlow(activity, task.result)

			// The flow has finished. The API does not indicate whether the user
			// reviewed or not, or even whether the review dialog was shown.
			//
			// No matter the result, just treat it as rated.
			flow.addOnCompleteListener { _ ->

				// Get the value indicating that it has been rated
				val rated = shared.resources.getInteger(R.integer.default_rate_my_app_rated)

				// Set the rated counter
				shared.rateMyAppCounter = rated

			}

		}
	}

}
